### PR TITLE
[gh-786] Add Ethereum RPC client API to client library.

### LINKED ISSUE #786: [rpc-client] add Ethereum RPC client to client library.
Part of #678

### PR DESCRIPTION
**Pull request by feliciss**
## Summary

Continuing after #787.

Add Ethereum RPC client API to client library.

- Closes #786

### CODE REVIEW

**Comment by baichuan3** [comment]
Next, I will refactor the key store to resolve https://github.com/rooch-network/rooch/issues/755. 
Before starting, please confirm that the refactor related to the key store in this PR will continue to be valid and will be merged into the main branch? @feliciss

**Comment by feliciss** [comment]
> Next, I will refactor the key store to resolve #755.
> Before starting, please confirm that the refactor related to the key store in this PR will continue to be valid and will be merged into the main branch? @feliciss

Not sure. I have a batch of commits that will be continuing after #788 to modify the key store.

Probably I'll modify these lines to change the generics:

https://github.com/rooch-network/rooch/blob/ac6a11cf8f2469c38e1c6a13dcbcdedde569a195/crates/rooch-key/src/keystore.rs#L565-L576

If it's not related to generics or it's not big changes, you can do refactors with it.

**Comment by baichuan3** [comment]
> Cuz

get it

**Comment by feliciss** [comment]
Can you check what the cause it is? May it relate to #842? @jolestar 

```cmd
✘  Then assert: "{{$.move[-1].execution_info.status.type}} == executed"
Step failed:
Defined: features/cmd.feature:59:7
Matched: crates/testsuite/tests/integration.rs:103:1
Step panicked. Captured output: expected 3 arguments: first [==|!=] second, but got input " == executed"
```

File: crates/rooch-key/src/keypair.rs [hunk 1]
```diff
@@ -28,8 +28,8 @@ use strum_macros::EnumString;
 #[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
 #[strum(serialize_all = "lowercase")]
 pub enum KeyPairType {
-    RoochKeyPairType,
-    EthereumKeyPairType,
+    Rooch,
+    Ethereum,
 }
```
Comment (jolestar) [inline crates/rooch-key/src/keypair.rs:33]: Make this rename to another PR, and only do the rename.

**Review by jolestar — COMMENTED** [review]
Use EthereumAddress to distinguish KeyStore, WalletContext and ClientConfig, It feels a little over-engineered and could lead to increased complexity.

**Comment by jolestar** [comment]
this PR only implements the Ethereum RPC client, do not change the KeyStore or WalletContext.

**Comment by feliciss** [comment]
Will remove key store refactoring and only keep Ethereum RPC client and add test cases.

File: crates/rooch-key/src/keypair.rs (hunk 1 above)
Comment (feliciss) [reply crates/rooch-key/src/keypair.rs]: I've moved it back.

**Comment by feliciss** [comment]
> > Can you check what the cause it is when running `cargo test -p testsuite --test integration`? May it relate to #842? @jolestar
> > ```batchfile
> > ✘  Then assert: "{{$.move[-1].execution_info.status.type}} == executed"
> > Step failed:
> > Defined: features/cmd.feature:59:7
> > Matched: crates/testsuite/tests/integration.rs:103:1
> > Step panicked. Captured output: expected 3 arguments: first [==|!=] second, but got input " == executed"
> > ```
> 
> Can you please figure out which command caused this error? It seems the last `rooch move` command is failed which is expected success. You can also run the failed command in you shell and see the detailed output.

This issue has been resolved.

File: crates/rooch/src/commands/transaction/commands/get_tx_by_hash_and_index.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+/// Get transaction by hash for Ethereum blockchain
+#[derive(Debug, Parser)]
+pub struct GetByHashAndIndexCommand {
+    /// Transaction's hash
+    #[clap(long)]
+    pub hash: H256,
+
+    /// Transaction's index
+    #[clap(long)]
+    pub index: u64,
```
Comment (baichuan3) [inline crates/rooch/src/commands/transaction/commands/get_tx_by_hash_and_index.rs:20]: `GetByHashAndIndexCommand ` has been removed in earlier versions.
This PR is far away from the main branch and needs to be rebased first for review.

Comment (feliciss) [reply crates/rooch/src/commands/transaction/commands/get_tx_by_hash_and_index.rs:20]: It's already in a rebased version, but I forgot removing them.

Please list all the possible removals. Sorry for the long version gap.

Comment (feliciss) [reply crates/rooch/src/commands/transaction/commands/get_tx_by_hash_and_index.rs]: I've removed `GetByHashAndIndexCommand`.

File: crates/testsuite/features/cmd.feature [hunk 3]
```diff
@@ -150,6 +150,6 @@ Feature: Rooch CLI integration tests
     Scenario: rpc test
       Given a server for rpc
       Then cmd: "rpc request --method eth_getBalance --params \"0x1111111111111111111111111111111111111111\""
-      Then assert: "{{$.rpc[-1]}} == 0x56bc75e2d63100000"
+      Then assert: "{{$.rpc.ethereum[-1]}}" equals "0x56bc75e2d63100000"
```
Comment (jolestar) [inline crates/testsuite/features/cmd.feature:153]: The response must be the same as the ETH RPC response.

https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_getbalance

```
// Request
curl -X POST --data '{"jsonrpc":"2.0","method":"eth_getBalance","params":["0x407d73d8a49eeb85d32cf465507dd71d507100c1", "latest"],"id":1}'
// Result
{
  "id":1,
  "jsonrpc": "2.0",
  "result": "0x0234c8a3397aab58" // 158972490234375000
}
```

**Review by baichuan3 — APPROVED** [review]

File: crates/rooch-rpc-api/src/jsonrpc_types/move_types.rs [hunk 4]
```diff
@@ -383,6 +384,78 @@ impl From<H256View> for H256 {
...
+    }
+}
+
+impl From<U256View> for U256 {
+    fn from(value: U256View) -> Self {
+        U256(value.0)
+    }
+}
+
+#[serde_as]
```
Comment (jolestar) [inline crates/rooch-rpc-api/src/jsonrpc_types/move_types.rs:443]: We use `StrView<u64>` in json for `u64`. Do we need a `U64View` for `U64` or also use `StrView<u64>`?

File: crates/rooch-rpc-api/src/jsonrpc_types/move_types.rs [hunk 5]
```diff
@@ -383,6 +384,78 @@ impl From<H256View> for H256 {
...
+
+impl From<H160View> for H160 {
+    fn from(value: H160View) -> Self {
+        H160(value.0)
+    }
+}
+
+#[serde_as]
+#[derive(Default, Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize, JsonSchema)]
+pub struct U256View([u64; 4]);
```
Comment (jolestar) [inline crates/rooch-rpc-api/src/jsonrpc_types/move_types.rs:429]: In Rooch API, we use `StrView<U256>` for U256. The `U256View` in JSON is a u64 Array?

File: crates/rooch-rpc-api/src/jsonrpc_types/move_types.rs [hunk 6]
```diff
@@ -383,6 +384,78 @@ impl From<H256View> for H256 {
...
+
+impl From<H64View> for H64 {
+    fn from(value: H64View) -> Self {
+        H64(value.0)
+    }
+}
+
+#[serde_as]
+#[derive(Default, Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize, JsonSchema)]
+pub struct H160View(
```
Comment (jolestar) [inline crates/rooch-rpc-api/src/jsonrpc_types/move_types.rs:409]: Is the `H160` in JSON need the `0x` prefix?

**Review by jolestar — CHANGES_REQUESTED** [review]
Need to make sure the response is compatible with Ethereum RPC and MetaMask.

File: crates/testsuite/features/cmd.feature (hunk 3 above)
Comment (jolestar) [reply crates/testsuite/features/cmd.feature:153]: Why the RPC response result contains an `ethereum` field?

Comment (feliciss) [reply crates/testsuite/features/cmd.feature:153]: I don't know where to handle this, but in the original code it had an error:

>    ✘  Then assert: "{{$.rpc}} == 0x56bc75e2d63100000"
      Step failed:
      Defined: features/cmd.feature:7:7
      Matched: crates/testsuite/tests/integration.rs:102:1
      Step panicked. Captured output: assertion failed: `(left == right)`
        left: `"[Error: RPC call failed: ErrorObject { code: MethodNotFound, message: \"Method not found\", data: None }]"`,
       right: `"0x56bc75e2d63100000"`: Assert "[Error: RPC call failed: ErrorObject { code: MethodNotFound, message: \"Method not found\", data: None }]" == "0x56bc75e2d63100000" failed

File: crates/rooch-rpc-api/src/jsonrpc_types/move_types.rs (hunk 6 above)
Comment (feliciss) [reply crates/rooch-rpc-api/src/jsonrpc_types/move_types.rs]: I have changed it to `H176View` since it's 22 bytes adding `0x` prefix.

File: crates/rooch-rpc-api/src/jsonrpc_types/move_types.rs (hunk 5 above)
Comment (feliciss) [reply crates/rooch-rpc-api/src/jsonrpc_types/move_types.rs]: Sorry. I didn't mean to use u64 at first place and forgot to change it back. I have fixed it now.

File: crates/rooch-rpc-api/src/jsonrpc_types/move_types.rs (hunk 4 above)
Comment (feliciss) [reply crates/rooch-rpc-api/src/jsonrpc_types/move_types.rs]: I changed it to `StrView<u64>`.

File: crates/testsuite/features/cmd.feature (hunk 3 above)
Comment (feliciss) [reply crates/testsuite/features/cmd.feature]: I changed it to `{{$.result}}`.

File: crates/rooch-rpc-api/src/jsonrpc_types/bytes.rs [hunk 7]
```diff
@@ -1,28 +1,43 @@
...
     }
 }
 
+impl FromStr for Bytes {
+    type Err = FromHexError;
+
+    /// Convert from a hexadecimal string representation of bytes.
+    fn from_str(hex_string: &str) -> Result<Self, FromHexError> {
+        // Use the `hex` crate to parse the hexadecimal string into bytes.
+        let bytes = hex::decode(hex_string)?;
```
[the remark this answers is not in the document]
Comment (feliciss) [reply crates/rooch-rpc-api/src/jsonrpc_types/bytes.rs:42]: I have stripped it off.

**Review by jolestar — APPROVED** [review]
